### DISCUSSION TITLE
trigger function to delete folder in storage

### DISCUSSION THREAD

**Discussion by pierroo** [supabase/supabase#16369] [Questions]
Hi,

I am trying to create a function that will be triggered upon a table entry deletion.
(that part I know how to manage)

However, I really found nothing online about storage function in plpgsql to delete stuff from storage?
![image](https://github.com/supabase/supabase/assets/17619412/c40fee6b-5b05-46fa-bc48-eb4a42f14c7c)

Is there actually anyway to deal with storage from a function?
And, to delete a folder entirely (not a bucket, not a specific file, but a folder containing many files at once)

**Comment by GaryAustin1** [comment]
You have to use pg_net or http extension to make storage api calls.
The basic framework guides in docs all  show an example of a trigger and using http extension to delete a storage object.
I also have an approach I show here: https://github.com/GaryAustin1/supa-file-helper

**Reply by pierroo** [reply]
Thank you @GaryAustin1 , I actually saw that link of yours in another post, but it just felt "overkill" to have to:
1- flag storage items for deletion from sql code
2- setup a cron task to run deletion over those flagged items

Rather than simply have a direct function to delete an entire **folder** at once, regardless of its content.

Otherwise, would it be possible to run a cron job to trigger an edge function to delete specific list of FOLDERS from storage?

**Reply by GaryAustin1** [reply]
You can certainly call an edge function using the http or pg_net extensions instead of the storage-api.  Not sure how that benefits you over just calling the storage-api from a database function based on cron.

**Reply by GaryAustin1** [reply]
Deleting a folder works out well as you can use bulk delete to speed it up.

**Reply by pierroo** [reply]
Just me being more familiar with node / edge function here to be honest.

And I also REALLY want to have it attached to a database trigger that seems so much cleaner for my use case.

Thank you!

**Reply by pierroo** [reply]
Bulk delete?
So I would still need to list all files in said folder first, then run a bulk delete through it?
No actual folder deletion function?

**Reply by GaryAustin1** [reply]
Correct, but only one api call.

**Comment by pierroo** [comment]
> Correct, but only one api call.

I'm sorry, I browsed for quite a while but I cannot find any "bulk delete" in one api call;
all the stuff I find online are about list then loop and delete one by one;
Can you please guide me here as well?

Edit: oh, I misunderstood, so you mean only one api call for the deletion itself; not for the listing + delete
I did see that delete in the folder can take an array indeed

**Reply by GaryAustin1** [reply]
![image](https://github.com/supabase/supabase/assets/54564956/200d7b37-eadc-4664-90b8-16ee5f9f2a5e)
https://supabase.com/docs/reference/javascript/storage-from-remove
Only works if all are in the bucket.  If you look at my repository I use bulk delete with both http and pg_net when I can.

**Reply by GaryAustin1** [reply]
Yes, listing if using the API would be needed.  If it is still in a database function you can just process storage.objects with SQL to get your pathnames and don't need the extra API call.
